### ISSUE TITLE
no-useless-escape false positive

### ISSUE THREAD

**Issue by thiagodelgado111** [eslint/eslint#7495]
**Tell us about your environment**

* **ESLint Version:** 3.9.0
* **Node Version:** node v6.9.1
* **npm Version:** 3.3.2

**What parser (default, Babel-ESLint, etc.) are you using?**
babel-eslint

**Please show your full configuration:**

```
{
  "extends": ["airbnb"],
  "plugins": ["mocha", "no-unused-vars-rest"],
  "env":{
    "meteor": true,
    "es6": true,
    "commonjs": true
  },
  "parserOptions": {
    "ecmaVersion": 6,
    "sourceType": "module"
  },
  "rules": {
    "consistent-return": [0],
    "eol-last": [0],
    "eqeqeq": ["error", "smart"],
    "max-len": ["error", 150, 2],
    "no-console": [0],
    "no-mixed-requires": [0],
    "no-native-reassign": [0],
    "no-param-reassign": [0],
    "no-use-before-define" : ["error", { "functions": false }],
    "no-useless-rename": ["error", {
      "ignoreDestructuring": true
    }],
    "no-undefined": ["error"],
    "no-underscore-dangle": ["error", { "allow": ["_id"], "allowAfterThis": true }],
    "object-curly-spacing": ["error", "always", {
      "objectsInObjects": false,
      "arraysInObjects": false
    }],
    "prefer-template": [0],
    "quotes": [2, "single"],
    "radix": ["error", "always"],
    "require-yield": [0],
    "semi": [2, "always"],
    "no-unused-vars": 0,
    "import/extensions": [0],
    "react/forbid-prop-types": [0],
    "jsx-a11y/no-static-element-interactions": [0],
    "arrow-parens":[2, "as-needed"],
    "react/no-unescaped-entities": [2],
    "react/no-unused-prop-types": [2],
    "no-plusplus":[2, { "allowForLoopAfterthoughts": true }],
    "no-bitwise": [2],
    "import/no-unresolved": [2, { ignore: ['\.sol\.js$'] }],
    "class-methods-use-this": [2],
    "jsx-a11y/anchor-has-content": [2],

    /* Plugins */
    "no-unused-vars-rest/no-unused-vars": [2, { "ignoreDestructuredVarsWithRest": true }],
    "babel/generator-star-spacing": [0],
    "generator-star-spacing": [0],
    "import/no-extraneous-dependencies": [0],
    "mocha/no-exclusive-tests": "error"
  },
  "settings": {
    "import/core-modules": ["meteor"],
    "import/resolver": "meteor"
  }
}
```

**What did you do? Please include the actual source code causing the issue.**

```js
const PATTERN_USERNAME = /^[a-z0-9\.\-_]+$/;
```

**What did you expect to happen?**
No errors

**What actually happened? Please include the actual, raw output from ESLint.**
`error  Unnecessary escape character: \.  no-useless-escape`

**Comment by not-an-aardvark** [comment]
Hi, thanks for the report. However, I think this is working as intended; `.` characters don't need to be escaped in regex character classes. To fix the issue, you can replace your code with this:

```js
const PATTERN_USERNAME = /^[a-z0-9.\-_]+$/;
```

...which will do exactly the same thing.

**Comment by thiagodelgado111** [comment]
The same is happening to other characters too `-, $, /, ^`

```
14:40   error  Unnecessary escape character: \$                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     no-useless-escape
  14:45   error  Unnecessary escape character: \*                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     no-useless-escape
  14:47   error  Unnecessary escape character: \+                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     no-useless-escape
  14:51   error  Unnecessary escape character: \/                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     no-useless-escape
  14:54   error  Unnecessary escape character: \?                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     no-useless-escape
  14:56   error  Unnecessary escape character: \^                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     no-useless-escape
  14:61   error  Unnecessary escape character: \|                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     no-useless-escape
  14:126  error  Unnecessary escape character: \$                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     no-useless-escape
  14:131  error  Unnecessary escape character: \*                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     no-useless-escape
  14:133  error  Unnecessary escape character: \+                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     no-useless-escape
  14:137  error  Unnecessary escape character: \/                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     no-useless-escape
  14:140  error  Unnecessary escape character: \?                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     no-useless-escape
  14:142  error  Unnecessary escape character: \^                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     no-useless-escape
  14:147  error  Unnecessary escape character: \|                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            no-useless-escape
```

**Comment by not-an-aardvark** [comment]
Yes, this is intended behavior; those characters do not need to be escaped in character classes. If you remove the backslashes before the unnecessary escape characters, your code should work exactly the same.

**Comment by thiagodelgado111** [comment]
Ok, so even being special chars, those doesn't need to be escaped depending on how we use them inside our expression, correct?

**Comment by not-an-aardvark** [comment]
Right. They only need to be escaped if they appear outside of character classes (the sections surrounded by `[]`).

```js
var foo = /\./; // needs to be escaped
var foo = /[\.]/; // does not need to be escaped
```

**Comment by thiagodelgado111** [comment]
Oh, Ok! Got it! I will close the issue then, thanks a lot for helping! :)
